core/src/main/java/hudson/model/Job.java

```diff
@@ -1143,6 +1148,22 @@ private HealthReport getBuildStabilityHealthReport() {
         int failCount = 0;
         int totalCount = 0;
         RunT i = getLastBuild();
+        RunT u = getLastFailedBuild();
+        if (i != null && u == null) {
+                // no failures, like ever
+                return new HealthReport(100, Messages._Job_BuildStability(Messages._Job_NoRecentBuildFailed()));
+        }
+        if (i != null && u.getNumber() <= i.getNumber()) {
+            int recentSuccessCount = _getRuns().subMap(u.getNumber() + 1, i.getNumber() + 1).size();
+            if (recentSuccessCount >= 5) {
+                return new HealthReport(100, Messages._Job_BuildStability(Messages._Job_NoRecentBuildFailed()));
+            } else {
+                // no need to load the intermediate builds, they're all successful, let's start iterating from
```

Comment:
Are you sure calling `subMap` and then `size` does not force build loading? Offhand I would guess that it would. You need a test case using `RunLoadCounter` to prove that your fix is effective (and that the added complexity of this block did not introduce regressions—the correctness of this code is hardly self-evident).